Syncing ldap group descriptions

beanaas opened the pull request:
Syncing group descriptions with LDAP.

https://github.com/zulip/zulip/issues/31040
https://chat.zulip.org/#narrow/stream/31-production-help/topic/sync.20group.20descriptions.20from.20ldap/near/1892521

Extended ZulipLDAPUser to have function for getting descriptions from LDAP. The function group_description_from_info is similar to that used in https://github.com/django-auth-ldap/django-auth-ldap/blob/master/django_auth_ldap/config.py for getting the group names. 

Group descriptions are now synced as part of the sync_groups_from_ldap process. This ensures that group descriptions for all relevant users are checked during synchronization. However, this method isn't the most efficient, as the descriptions only need to be synced once. Optimizing this would require either modifying the existing sync_groups_from_ldap command or creating a new command specifically for syncing group descriptions.

A new setting, LDAP_GROUP_DESCRIPTION_ATTR, has been introduced to map descriptions to an attribute in an LDAP group. To enable description syncing, this setting must be configured. If left unset, no syncing will occur, meaning organizations that do not explicitly enable this feature will not experience any changes.

TODO:
Documentation of  description syncing.

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

beanaas commented on zproject/backends.py:
```diff
@@ -622,6 +623,39 @@ def email_belongs_to_ldap(realm: Realm, email: str) -> bool:
+            name = ci_dict["cn"][0]
+            description = ci_dict.get(GROUP_DESCRIPTION_ATTR, None)
+            group_description[name] = description
+        return group_description
```
So something like this?

```
class ZulipLDAPUser(_LDAPUser):
    """
    This is an extension of the _LDAPUser class, with a realm attribute
    attached to it. It's purpose is to call its inherited method
    populate_user() which will sync the LDAP data with the corresponding
    UserProfile. The realm attribute serves to uniquely identify the UserProfile
    in case the LDAP user is registered to multiple realms.
    """

    def __init__(self, *args: Any, **kwargs: Any) -> None:
        self.realm: Realm = kwargs["realm"]
        del kwargs["realm"]

        super().__init__(*args, **kwargs)

    def _get_group_descriptions(self):
        if self._groups is None:                                                                        
            self._groups = _LDAPUserGroups(self)
        group_infos = self._groups._get_group_infos()
        ldap_group_descriptions ={}
        for group_info in group_infos:
            group_name = self._groups._group_type.group_name_from_info(group_info)
            group_description = self.group_description_from_info(group_info)
            ldap_group_descriptions[group_name] = group_description                                                               
        return ldap_group_descriptions 
    
    def group_description_from_info(self, group_info):
        try:
            description = group_info[1][settings.LDAP_GROUP_DESCRIPTION_ATTR]
        except (KeyError, IndexError):
            description = None
        return description
```

or should the ZulipLDAPUser have additional field for storing the descriptions?

timabbott replied:
Yeah something like that!

timabbott commented:
@mateuszmandera can you review this one at this point?

mateuszmandera reviewed (commented):
Thanks for working on this, looks close to being good to go already! I posted several minor comments on things we'll want to tweak.

beanaas commented on zerver/tests/test_auth_backends.py:
```diff
@@ -7999,6 +7999,53 @@ def test_ldap_group_sync(self) -> None:
                 'WARNING:django_auth_ldap:search_s("ou=groups,dc=zulip,dc=com", 1, "(&(objectClass=groupOfUniqueNames(uniqueMember=uid=cordelia,ou=users,dc=zulip,dc=com))", "None", 0) while authenticating cordelia',
             ],
         )
+        with (
```
Added a comment on top, folding it into the code above is possible but it involved changing the assertions quite much since sync ldap descriptions will be called multiple times. Could do it if it is something you really want

mateuszmandera replied:
Yeah a comment is fine i think.

beanaas committed:
ldap: Add support for syncing of group descriptions.

Fixes #31040.

Co-authored-by: Mateusz Mandera <mateusz.mandera@zulip.com>

mateuszmandera reviewed (commented):
Thanks. looks pretty good! I made some additional tweaks and pushed here.

mateuszmandera commented on zerver/tests/test_auth_backends.py:
```diff
@@ -6762,6 +6763,109 @@ def test_login_success_when_user_does_not_exist_with_split_full_name_mapping(sel
+            )
+
+    @override_settings(AUTHENTICATION_BACKENDS=("zproject.backends.ZulipLDAPAuthBackend",))
+    def test_get_group_descriptions_when_ldap_groups_exist_when_LDAP_GROUP_DESCRIPTION_ATTR_not_defined(
```
I'll resolve this since this is fixed in my tweaks I pushed here.

mateuszmandera commented:
So I'd just like some confirmation on the two notes/questions I left above. After that I can add some documentation and then should be ready for integration review.

beanaas commented on zerver/tests/fixtures/ldap/directory.json:
```diff
@@ -87,14 +87,16 @@
             "uid=cordelia,ou=users,dc=zulip,dc=com"
-        ]
+        ],
+        "description": "another_test_group_description"
```
No I have not tested it with a real LDAP. You are absolutely correct I misunderstood RFC4519, it will be returned in a list. Thanks for pointing that out, the way you showed 
```
description = group_info[1][settings.LDAP_GROUP_DESCRIPTION_ATTR][0]
```

Do you want me to correct this?

mateuszmandera replied:
@beanaas Yup, that would be helpful! It's also pretty important to give this some basic testing with a real LDAP setup, just to be sure end-to-end that this works. Are you able to do such a test? If you have your own self-hosted server where you're losing LDAP, that'd probably be pretty easy by just deploying this pull request there and testing. Otherwise it involves setting up some kind of ldap integration in the development environment. 

If you won't be able to test that, I'll try to find some time to do the development environment set up myself and give this a try; because I think we need to do that before merging.

beanaas replied:
Okey, will fix that. I don't have a real LDAP setup, but I will try to create one and test on that. If I can't manage to fix that I will write to you.

mateuszmandera replied:
Okay sounds good - thank you!

beanaas replied:
@mateuszmandera Now I have tested with a real LDAP server and it worked as you said, sorry for the late response took more time then anticipated setting the LDAP up.

beanaas commented on zproject/backends.py:
```diff
@@ -1194,6 +1224,28 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
+            return {}
+        ldap_group_descriptions: dict[str, Any] = {}
+        self._groups: _LDAPUserGroups = self._get_groups()
+        group_infos: list[tuple[str, dict[str | None, Any]]] = self._groups._get_group_infos()
```
None should not be a dict key, so your approach seems correct!

beanaas commented:
@mateuszmandera

mateuszmandera commented on docs/production/authentication-methods.md:
````diff
@@ -264,6 +264,24 @@ groups. To configure this feature:
    and `group2` will match their membership in LDAP groups with those
    names.
 
+1. Optionally, you can also enable syncing of descriptions of the
+   configured groups. The relevant Zulip groups will have their
+   descriptions updated to match the descriptions in LDAP.
+   To enable this, simply set:
+
+   ```python
+   LDAP_GROUP_DESCRIPTION_ATTR = "description"
+   ```
+
+   in place of `"description"` using the actual attribute name of your
+   LDAP groups that should be synced.
+   If a particular group doesn't have this attribute in your LDAP
+   directory, the Zulip group description will remain unchanged.
+
+   Note that this is a server-level setting, so specifying a different
+   LDAP attribute for the group description for different realms
+   is not supported.
````
@beanaas I fixed a typo and added a bit of documentation. Can you just double-check it quickly and confirm this sounds correct?

beanaas replied:
@mateuszmandera yes that sounds correct!

mateuszmandera commented:
Cool, tagged for `integration review` now. @timabbott Can you review next? The configuration interface via the `LDAP_GROUP_DESCRIPTION_ATTR` server-level setting might be worth some consideration -  but it makes sense to me.

It might be a bit weird that you can configure the group lists per-realm, but this is only server-level - but it seems even stranger and unlikely to have different LDAP attribute names for group description for different realms... So while we could allow customizing this per-realm, it seems like an excessive level of granularity.

timabbott commented on zproject/backends.py:
```diff
@@ -1018,6 +1021,38 @@ def sync_groups_from_ldap(self, user_profile: UserProfile, ldap_user: _LDAPUser)
+                        )
+                        do_update_user_group_description(
+                            user_group, ldap_description, acting_user=None
+                        )
```
It looks like this does this loop for each user being synced, which is a lot of useless duplicate work. I think we need this to be a new top-level function, called directly from `sync_ldap_user_data` outside the loop.

mateuszmandera replied:
@timabbott This description-sync indeed gets run per user, but I don't think it's that much duplication? But the worst thing that gets run like is the group fetch
```
            zulip_group_descriptions_dict = {
                group.name: group.description
                for group in NamedUserGroup.objects.filter(
                    realm=user_profile.realm, name__in=configured_ldap_group_names_for_sync
                )
            }
```

which is probably not a big? The actual updates don't get duplicated because the loop checks the Zulip group description need to be updated. So any group which needs updating, will get updated once and that's all.

I agree that the natural way for this functionality to work would be outside of the users loop, as a separate sync ldap task which just:
1. Fetches all groups in LDAP with the description values.
2. Fetches all ldap-syncable Zulip groups
3. Does the appropriate sync.

But I don't know if `django-auth-ldap` provides an api for doing (1). Maybe i missed something obvious, but didn't find. Because it's focused on being a user auth library, its view of operating on groups is  "groups this specific user belongs to". So we actually have to work in a per-user loop because we can only fetch groups per-user.

timabbott replied:
I think if you've got 10K users and 25 groups, it seems bad to do fetch the 25 groups 10K times. Probably we can just have a `already_synced_groups` set that keeps track of which ones we've processed in this run?

mateuszmandera replied:
@timabbott  We don't have a very good way to do that, this is called inside of `ZulipLDAPUserPopulator.get_or_build_user`, which is called inside `django-auth-ldap` - so we don't have the control to pass state like this from iteration to iteration. We could probably work out something very hacky  like tacking on the `.already_synced_groups` attribute on the `ZulipLDAPUser` object being synced  - which could then be read from inside `get_or_build_user`. 

So that sounds ugly enough that I left it out for now. Just made the query a bit better instead:
```
            zulip_group_descriptions_dict = {
                group.name: group.description
                for group in NamedUserGroup.objects.filter(
                    # We only need to fetch the groups for which the user is a membr
                    # of the corresponding LDAP group. We don't have ldap description data for other
                    # groups, so they won't be subject to syncing anyway.
                    realm=user_profile.realm,
                    name__in=ldap_group_name_set_for_user,
                ).only("name", "description")
            }
```

timabbott replied:
I was thinking of just a global cache, that could be passed into this code path from the management command?

timabbott commented:
Noted one fairly serious bug, but otherwise this looks good (though I'll do a final copy-editing pass on the docs before merging).

timabbott commented:
>  The configuration interface via the LDAP_GROUP_DESCRIPTION_ATTR server-level setting might be worth some consideration - but it makes sense to me.
> 
> It might be a bit weird that you can configure group lists per-realm, but this is only server-level - but it seems even stranger and unlikely to have different LDAP attribute names for group description for different realms... So while we could allow customizing this per-realm, it seems like an excessive level of granularity.

Yeah, I think this is correct; we will want the entire set of LDAP configuration variables to be per-realm.

mateuszmandera commented:
> Yeah, I think this is correct; we will want the entire set of LDAP configuration variables to be per-realm.

Hmm, Is your point that we want to have this as this global-style variable for now, because once we transition to a proper per-realm ldap configuration framework, settings of this kind will naturally get transitioned to be per-realm? That's my thinking too, if I read this correctly.

timabbott commented:
Yep, exactly.

mateuszmandera commented:
@timabbott  Replied in https://github.com/zulip/zulip/pull/31863#discussion_r1897579970 about why storing the `already_synced_groups` would have to be very ugly/hacky and just pushed a commit improving that query + another commit to just clean up the `sync_groups_from_ldap` function overall to read cleaner.

mateuszmandera committed:
sync_groups_from_ldap: Keep a cache of groups which already got synced.

If a group gets it description synced during a sync_user_from_ldap call
for a specific user, it doesn't need to repeatedly get re-fetched during
the processing of other users.
This allows us to avoid a lot of unnecessary fetching from db if there
are many group memberships per-user that we're reading from LDAP.

mateuszmandera commented:
I resolved the conflicts. Tagging for `integration review`, since I think this is in a pretty decent state, pending decision on whether we want something like the second commit for optimization.
